streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -596,6 +596,68 @@ class ZStream[-R, +E, +A](val process: ZManaged[R, E, Pull[R, E, A]]) extends Se
       ZStream.fromQueue(queue).unTake
     }
 
+  /**
+   * Creates a stream that passes its elements through a queue that drops elements
+   * (dropping or sliding). It implements a promise that signals abrupt completion
+   * of the stream due to an error or the end of the stream. After the promise has
+   * been completed, it continues to offer the current (at the time) contents of the
+   * queue, until it has been exhausted, at which point it finishes with the signal
+   * from the promise.
+   */
+  private final def bufferSignal[E1 >: E, A1 >: A](queue: Queue[Take[E1, A1]]): ZStream[R, E1, A1] =
+    ZStream.managed {
+      for {
+        signal <- Promise.make[Nothing, Take[E, Nothing]].toManaged_
```

Comment:
I think you can get rid of `Take` entirely here by emitting plain values in the queue and making this promise a `Promise[Option[E], Nothing]`. Unifying the two gives you `Pull`.

Reply:
What about the `Cause`?